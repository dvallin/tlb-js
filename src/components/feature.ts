import { Color } from '../renderer/color'
import { gray, primary } from '../renderer/palettes'
import { TlbWorld } from '../tlb'

import { FeatureComponent } from './feature'
import { PositionComponent } from './position'
import { WorldMap } from '../resources/world-map'
import { Vector } from '../spatial'
import { Entity } from '../ecs/entity'
import { strangeSymbols } from '../symbols'

export type FeatureType = keyof typeof features
export interface FeatureComponent {
  type: FeatureType
}

export interface Feature {
  character: string
  diffuse: Color

  blocking: boolean
  lightBlocking: boolean
  description: string
}

export const features = {
  wall: {
    character: '#',
    diffuse: gray[3],
    blocking: true,
    lightBlocking: true,
    description: 'a wall',
  },
  corridor: {
    character: '.',
    diffuse: gray[0],
    blocking: false,
    lightBlocking: false,
    description: 'a corridor',
  },
  room: {
    character: '.',
    diffuse: primary[1],
    blocking: false,
    lightBlocking: false,
    description: 'floor of a room',
  },
  locker: {
    character: strangeSymbols[16],
    diffuse: primary[1],
    blocking: true,
    lightBlocking: true,
    description: 'a locker',
  },
  trash: {
    character: strangeSymbols[21],
    diffuse: gray[2],
    blocking: false,
    lightBlocking: false,
    description: 'some trash',
  },
  door: {
    character: strangeSymbols[27],
    diffuse: primary[1],
    blocking: true,
    lightBlocking: true,
    description: 'a door',
  },
  hub: {
    character: '.',
    diffuse: primary[0],
    blocking: false,
    lightBlocking: false,
    description: 'floor of a hub',
  },
  player: {
    character: '@',
    diffuse: primary[0],
    blocking: true,
    lightBlocking: true,
    description: 'you',
  },
  guard: enemy('g', 'a guard'),
  eliteGuard: eliteEnemy('g', 'a very strong guard'),
}

function enemy(character: string, description: string): Feature {
  return {
    character,
    diffuse: primary[1],
    blocking: true,
    lightBlocking: true,
    description,
  }
}

function eliteEnemy(character: string, description: string): Feature {
  return {
    character,
    diffuse: primary[3],
    blocking: true,
    lightBlocking: true,
    description,
  }
}

export function getFeature(world: TlbWorld, entity: number): Feature | undefined {
  const feature = world.getComponent<FeatureComponent>(entity, 'feature')
  if (feature) {
    return features[feature.type]
  }
  return undefined
}

export function createFeature(world: TlbWorld, map: WorldMap, position: Vector, type: FeatureType): Entity {
  const e = map.getTile(position)
  if (e !== undefined) {
    throw new Error(`there is already a tile at ${position.key}`)
  }

  const entity = world
    .createEntity()
    .withComponent<PositionComponent>('position', { position })
    .withComponent<FeatureComponent>('feature', { type }).entity
  map.setTile(position, entity)
  return entity
}
