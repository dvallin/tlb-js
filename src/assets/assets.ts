import { Asset, Cover } from '../components/asset'
import { Vector } from '../spatial'
import { features, generators } from './features'
import { Feature } from '../components/feature'

const s1x1 = new Vector([1, 1])
const s1x2 = new Vector([1, 1])
const s3x1 = new Vector([3, 1])
const s2x2 = new Vector([2, 2])
function asset(name: string, size: Vector, cover: Cover, feature: (index: number) => Feature): Asset {
  return { name, size, cover, hasInventory: true, feature }
}

const assetsDefinition = {
  door: asset('metal door', s3x1, 'full', () => features['door']),
  locker: asset('a locker', s1x1, 'full', () => features['locker']),
  trash: asset('some trash', s1x1, 'none', () => features['trash']),
  loot: asset('a dead body', s1x1, 'none', () => features['loot']),
  table: asset('a table', s1x2, 'partial', () => features['table']),
  generator: asset('a generator', s2x2, 'full', generators['block']),
}

export type AssetType = keyof typeof assetsDefinition
export const assets: { [key in AssetType]: Asset } = assetsDefinition
