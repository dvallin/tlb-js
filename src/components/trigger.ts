import { Entity } from '../ecs/entity'

export interface TriggersComponent {
  type: 'asset' | 'dialog'
  name: string
  entities: Entity[]
}

export interface TriggeredByComponent {
  entity: Entity
}
