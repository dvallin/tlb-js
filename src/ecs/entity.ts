import { World } from "./world"

export type Entity = number

export class EntityModifier<C, S, R> {

    public constructor(
        private readonly world: World<C, S, R>,
        public readonly entity: Entity
    ) { }

    public withComponent<T>(name: C, component: T): EntityModifier<C, S, R> {
        this.world.getStorage(name).insert(this.entity, component)
        return this
    }

    public removeComponent(name: C): EntityModifier<C, S, R> {
        this.world.getStorage(name).remove(this.entity)
        return this
    }

    public delete(): void {
        this.world.components.forEach(s => s.remove(this.entity))
    }
}
