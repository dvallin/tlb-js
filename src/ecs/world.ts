import { Storage } from './storage'
import { System } from './system'
import { Resource } from './resource'
import { Entity, EntityModifier } from './entity'

export class World<C, S, R> {
  public readonly components: Map<C, Storage<object>> = new Map()
  public readonly resources: Map<R, Resource<C, S, R>> = new Map()
  public readonly systems: Map<S, System<C, S, R>> = new Map()
  public readonly emptySystems: Set<S> = new Set()
  public readonly activeSystems: Set<S> = new Set()

  private openEntities: Set<Entity> = new Set()
  private lastEntity: Entity = -1

  public get entityCount(): number {
    return this.lastEntity + 1 - this.openEntities.size
  }

  public registerComponentStorage<T extends object>(component: C, storage: Storage<T>): void {
    this.components.set(component, storage)
  }

  public getStorage<T extends object>(component: C): Storage<T> {
    return this.components.get(component)! as Storage<T>
  }

  public getResource<T extends Resource<C, S, R>>(resource: R): T {
    return this.resources.get(resource)! as T
  }

  public getComponent<T extends object>(entity: Entity, component: C): T | undefined {
    const storage = this.getStorage<T>(component)
    if (storage !== undefined) {
      return storage.get(entity)
    }
    return undefined
  }

  public hasEntity(entity: Entity): boolean {
    return entity <= this.lastEntity && !this.openEntities.has(entity)
  }

  public hasComponent(entity: Entity, component: C): boolean {
    const storage = this.getStorage(component)
    if (storage !== undefined) {
      return storage.has(entity)
    }
    return false
  }

  public createEntity(): EntityModifier<C, S, R> {
    let entity
    if (this.openEntities.size > 0) {
      entity = this.openEntities.values().next().value
      this.openEntities.delete(entity)
    } else {
      entity = ++this.lastEntity
    }
    return new EntityModifier(this, entity)
  }

  public editEntity(entity: Entity): EntityModifier<C, S, R> {
    return new EntityModifier(this, entity)
  }

  public deleteEntity(entity: Entity): void {
    this.editEntity(entity).delete()
    this.openEntities.add(entity)
  }

  public registerSystem(name: S, system: System<C, S, R>): void {
    this.systems.set(name, system)
  }

  public enableSystem(name: S): void {
    this.activeSystems.add(name)
  }

  public disableSystem(name: S): void {
    this.activeSystems.delete(name)
  }

  public activeSystemsList(): S[] {
    const l: S[] = []
    this.activeSystems.forEach(s => l.push(s))
    return l
  }

  public registerResource(resource: Resource<C, S, R>): void {
    this.resources.set(resource.kind, resource)
  }

  public updateResources(): void {
    for (const resource of this.resources) {
      resource[1].update(this)
    }
  }

  public updateSystems(): void {
    this.emptySystems.clear()
    for (const name of this.activeSystems) {
      const system = this.systems.get(name)!
      const firstStorage = this.getStorage(system.components[0])
      let calls = 0
      firstStorage.foreach((entity: Entity) => {
        let callSystem = true
        for (let i = 1; i < system.components.length; i++) {
          if (!this.hasComponent(entity, system.components[i])) {
            callSystem = false
            break
          }
        }
        if (callSystem) {
          system.update(this, entity)
          calls++
        }
      })
      if (calls === 0) {
        this.emptySystems.add(name)
      }
    }
  }
}
