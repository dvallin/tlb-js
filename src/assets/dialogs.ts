import { TlbWorld } from '../tlb'
import { Entity } from '../ecs/entity'
import { InventoryComponent, ItemComponent } from '../components/items'
import { ItemType } from './items'
import { TriggeredByComponent, TriggersComponent } from '../components/trigger'
import { DialogComponent } from '../components/dialog'

export type AnswerType = 'close' | 'attack' | 'authorized'
export type Check = (world: TlbWorld, player: Entity, npc: Entity) => boolean
export interface Answer {
  text: string
  navigation: number | AnswerType
  check?: Check
}
export interface Step {
  text: string[]
  answers: Answer[]
}
export interface Dialog {
  steps: Step[]
}

function answer(text: string, navigation: number | AnswerType, check?: Check): Answer {
  return { check, text, navigation }
}
function attack(): Answer {
  return answer('[attack]', 'attack')
}
const playerHasItem = (type: ItemType) => (world: TlbWorld, player: Entity) => {
  const inventory = world.getComponent<InventoryComponent>(player, 'inventory')!.content
  return inventory.some(i => world.getComponent<ItemComponent>(i, 'item')!.type === type)
}

const dialogDefinitions = {
  randomRemarks: {
    steps: [
      {
        text: [
          'we suspect that your sister controls the mutant rom chip and the frozen wand.',
          'I apoligize for this statement',
          'The CNT-FAI briefly considered joining the Comintern because they fetishized revolutionary upsurges and apparently knew nothing about politics or socialism',
          'Thinking about building a railgun',
          'Kinda want to hug, but like irl. for the propriocetive stimmy feels :P',
          'I am, at my core, a stubborn and frustrating human who only exists to be set in my ways.',
          "England are gonna win. And I'm gonna post so many garbage selfies tonight",
          'Love that my old man is spending his saturday swearing at remainers appearing on the news',
          "you think you do but you don't",
          'I reject nation states',
          'Woah I wish I could read',
          "What the hell is 'pajama'",
          'Always use protection',
          'Subsurface warfare is going to get lit',
          'Gracious wife, gracious life',
          'im about to nut these loads',
          'May biqueerplatonic people find success in their lives',
          'I hope hologender people are getting enough time to rest!',
          'Washing my chin',
          'I am still against full luxury automated communism. thats not what i aim for.',
          '#Javascript was a mistake.',
          'we had fun looking at all the medieval babies at the art gallery today',
          "Oh, me? I'm strictly chillin'",
          "It's a comfortable dusk in the city, and you are a comfortable salmon.",
          'adulthood is having your debit card declined, fashionably',
        ],
        answers: [answer('ok', 'close')],
      },
    ],
  },
  guardRandomRemarks: {
    steps: [
      {
        text: [
          'Do you want to cry now?',
          'Love a neurogender bitch, oh, it get my dick hard',
          "You ain't gonna let me fuck you and I feel you",
          "I'm icy bitch, don't look at my wrist.",
          'Fuck love. All I got for hoes is hard dick and bubblegum.',
          'I punched this transhumanist in the ripcage and kicked her in the stomach',
        ],
        answers: [answer('ok', 'close'), attack()],
      },
    ],
  },
  restrictedAreaCheck: {
    steps: [
      {
        text: ['Hey you, stop it right there. This is a restricted area.', 'Identify yourself'],
        answers: [attack(), answer('[show id card]', 1, playerHasItem('idCard'))],
      },
      {
        text: ['Ok, you can pass'],
        answers: [answer('ok', 'authorized')],
      },
    ],
  },
}
export type DialogType = keyof typeof dialogDefinitions
export const dialogs: { [key in DialogType]: Dialog } = dialogDefinitions

export function addDialog(world: TlbWorld, entity: Entity, type: DialogType): Entity {
  const dialog = world
    .createEntity()
    .withComponent<TriggeredByComponent>('triggered-by', { entity })
    .withComponent<DialogComponent>('dialog', { type }).entity
  world.getComponent<TriggersComponent>(entity, 'triggers')!.entities.push(dialog)
  return dialog
}
