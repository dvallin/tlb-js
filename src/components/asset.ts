import { Vector } from '../spatial'
import { TlbWorld } from '../tlb'
import { Entity } from '../ecs/entity'
import { WorldMap, WorldMapResource } from '../resources/world-map'
import { Shape } from '../geometry/shape'

import { FeatureComponent, createFeature, Feature, FeatureProvider } from './feature'
import { GroundComponent } from './ground'
import { PositionComponent } from './position'
import { Rectangle } from '../geometry/rectangle'
import { TriggersComponent } from './trigger'

import { AssetType, assets } from '../assets/assets'
import { FeatureType } from '../assets/features'
import { Direction } from '../spatial/direction'
import { InventoryComponent } from './items'
import { DialogType } from '../assets/dialogs'
import { DialogComponent } from './dialog'

export interface AssetComponent {
  type: AssetType
}

export interface Asset {
  name: string
  size: Vector
  hasInventory: boolean
  dialog: DialogType | undefined
  feature: (index: number) => Feature
}

export interface PlacedAsset {
  type: AssetType
  tiles: { position: Vector; feature: FeatureType }[]
}

export function shapeOfAsset(type: AssetType, position: Vector, direction: Direction): Shape {
  const asset = assets[type]
  return Rectangle.footprint(position, direction, asset.size)
}

export function createAsset(world: TlbWorld, level: number, position: Vector, direction: Direction, type: AssetType): Entity {
  const shape = shapeOfAsset(type, position, direction)
  return createAssetFromShape(world, level, shape, type)
}

export function createAssetFromShape(world: TlbWorld, level: number, shape: Shape, type: AssetType): Entity {
  const entity = placeAsset(world, type)
  placeAssetParts(world, type, entity, level, shape)
  return entity
}

function placeAsset(world: TlbWorld, type: AssetType): Entity {
  const asset = assets[type]
  const entity = world
    .createEntity()
    .withComponent<AssetComponent>('asset', { type })
    .withComponent<TriggersComponent>('triggers', { name: asset.name, type: 'asset', entities: [] }).entity
  if (asset.hasInventory) {
    world.editEntity(entity).withComponent<InventoryComponent>('inventory', { content: [] })
  }
  if (asset.dialog !== undefined) {
    world.editEntity(entity).withComponent<DialogComponent>('dialog', { type: asset.dialog })
  }
  return entity
}

function placeAssetParts(world: TlbWorld, type: AssetType, entity: Entity, level: number, shape: Shape): void {
  const map: WorldMap = world.getResource<WorldMapResource>('map')
  const asset = assets[type]
  shape.foreach((p, i) => {
    const tile = putTile(world, map, level, p, () => asset.feature(i))
    addTrigger(world, tile, entity)
  })
}

export function removeAsset(world: TlbWorld, map: WorldMap, entity: Entity): void {
  const triggers = world.getComponent<TriggersComponent>(entity, 'triggers')!
  triggers.entities.forEach(tile => removeTile(world, map, tile))
  world.deleteEntity(entity)
}

function removeTile(world: TlbWorld, map: WorldMap, entity: Entity): void {
  const ground = world.getComponent<GroundComponent>(entity, 'ground')!
  const position = world.getComponent<PositionComponent>(entity, 'position')!
  map.levels[position.level].removeTile(position.position)
  createFeature(world, position.level, position.position, ground.feature)
  world.deleteEntity(entity)
}

function putTile(world: TlbWorld, map: WorldMap, level: number, position: Vector, feature: FeatureProvider): Entity {
  const ground = removeGround(world, map, level, position)
  const tile = world
    .createEntity()
    .withComponent<PositionComponent>('position', { level, position })
    .withComponent<FeatureComponent>('feature', { feature })
    .withComponent<GroundComponent>('ground', { feature: ground }).entity
  map.levels[level].setTile(position, tile)
  return tile
}

function addTrigger(world: TlbWorld, tile: Entity, entity: Entity): void {
  world.editEntity(tile).withComponent('triggered-by', { entity })
  world.getComponent<TriggersComponent>(entity, 'triggers')!.entities.push(tile)
}

function removeGround(world: TlbWorld, map: WorldMap, level: number, position: Vector): FeatureProvider {
  const entity = map.levels[level].removeTile(position)
  if (entity === undefined) {
    throw new Error(`cannot build asset on missing ground at ${position.key}`)
  }
  const feature = world.getComponent<FeatureComponent>(entity, 'feature')
  if (feature === undefined || feature.feature().blocking) {
    throw new Error('cannot build asset on missing or blocking ground')
  }
  world.deleteEntity(entity)
  return feature.feature
}
