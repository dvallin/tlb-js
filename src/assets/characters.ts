import { BodyPart, CharacterStats, CharacterStatsComponent, createCharacterStatsComponent } from '../components/character-stats'
import { TlbWorld } from '../tlb'
import { features, FeatureType } from './features'
import { FeatureComponent } from '../components/feature'
import { FovComponent } from '../components/fov'
import { HasActionComponent } from '../components/action'
import { ItemComponent, InventoryComponent, EquipedItemsComponent } from '../components/items'
import { ActiveEffectsComponent } from '../components/effects'
import { CharacterCreator } from '../generative/complex-embedder'
import { Entity } from '../ecs/entity'
import { AiComponent } from '../components/ai'
import { ActionType } from './actions'
import { ItemType } from './items'
import { addDialog } from './dialogs'
import { TriggersComponent } from '../components/trigger'

function humanoidBodyParts(core: number, head: number, limp: number): { [key: string]: BodyPart } {
  return {
    head: { type: 'head', itemAttachments: ['helmet'], health: head },
    leftArm: { type: 'arm', itemAttachments: ['glove', 'weapon'], health: limp },
    rightArm: { type: 'arm', itemAttachments: ['glove', 'weapon'], health: limp },
    torso: { type: 'torso', itemAttachments: ['armor'], health: core },
    leftLeg: { type: 'leg', itemAttachments: ['pants', 'boots'], health: limp },
    rightLeg: { type: 'leg', itemAttachments: ['pants', 'boots'], health: limp },
  }
}

const charactersStatsDefinition = {
  player: {
    bodyParts: humanoidBodyParts(14, 5, 5),
    strength: 5,
    movement: 4,
    actions: 4,
    aim: 10,
  },
  guard: {
    bodyParts: humanoidBodyParts(3, 1, 1),
    strength: 5,
    movement: 3,
    actions: 3,
    aim: 6,
  },
  eliteGuard: {
    bodyParts: humanoidBodyParts(5, 3, 3),
    strength: 8,
    movement: 2,
    actions: 5,
    aim: 6,
  },
  boss: {
    bodyParts: humanoidBodyParts(8, 5, 5),
    strength: 10,
    movement: 3,
    actions: 9,
    aim: 8,
  },
}
export type CharacterStatsType = keyof typeof charactersStatsDefinition
export const characterStats: { [key in CharacterStatsType]: CharacterStats } = charactersStatsDefinition

const characterCreatorsDefinition = {
  player: (world: TlbWorld) => createPlayer(world),
  eliteGuard: (world: TlbWorld) => {
    const entity = createDefaultNpc(world, 'some elite guard', 'eliteGuard', 'eliteGuard')
    addDialog(world, entity, 'guardRandomRemarks')
    return entity
  },
  guard: (world: TlbWorld) => {
    const entity = createDefaultNpc(world, 'some guard', 'guard', 'guard')
    addDialog(world, entity, 'guardRandomRemarks')
    return entity
  },
}

export const defaultActions: ActionType[] = ['longMove', 'hit', 'rush', 'endTurn']

export function createPlayer(world: TlbWorld): Entity {
  const player = createCharacter(world, 'player', 'player', defaultActions)
  world
    .editEntity(player)
    .withComponent<{}>('player', {})
    .withComponent<FovComponent>('fov', { fov: [] })
    .withComponent<{}>('viewport-focus', {})

  take(world, player, 'idCard')
  take(world, player, 'rifle')
  take(world, player, 'sniperRifle')
  take(world, player, 'bootsOfStriding')
  equip(world, player, 'nailGun', ['leftArm'])
  equip(world, player, 'leatherJacket', ['torso', 'leftArm', 'rightArm'])
  return player
}

export function createDefaultNpc(world: TlbWorld, name: string, statsType: CharacterStatsType, featureType: FeatureType): Entity {
  const npc = createEmptyNpc(world, name, statsType, featureType, defaultActions)
  equip(world, npc, 'rifle', ['leftArm'])
  return npc
}

export function take(world: TlbWorld, entity: Entity, type: ItemType): Entity {
  const item = world.createEntity().withComponent<ItemComponent>('item', { type }).entity
  world.getComponent<InventoryComponent>(entity, 'inventory')!.content.push(item)
  return item
}

export function equip(world: TlbWorld, entity: Entity, type: ItemType, bodyParts: string[]): void {
  const item = take(world, entity, type)
  world.getComponent<EquipedItemsComponent>(entity, 'equiped-items')!.equipment.push({
    entity: item,
    bodyParts,
  })
}

export function createEmptyNpc(
  world: TlbWorld,
  name: string,
  statsType: CharacterStatsType,
  featureType: FeatureType,
  actions: ActionType[]
): Entity {
  const character = createCharacter(world, statsType, featureType, actions)
  return world
    .editEntity(character)
    .withComponent('npc', {})
    .withComponent<TriggersComponent>('triggers', { entities: [], type: 'dialog', name })
    .withComponent<AiComponent>('ai', { type: 'rushing', state: 'idle', interest: undefined, distrust: 0 }).entity
}

export function createCharacter(world: TlbWorld, statsType: CharacterStatsType, featureType: FeatureType, actions: ActionType[]): Entity {
  return world
    .createEntity()
    .withComponent<HasActionComponent>('has-action', { actions })
    .withComponent<InventoryComponent>('inventory', { content: [] })
    .withComponent<EquipedItemsComponent>('equiped-items', { equipment: [] })
    .withComponent<FeatureComponent>('feature', { feature: () => features[featureType] })
    .withComponent<CharacterStatsComponent>('character-stats', createCharacterStatsComponent(statsType))
    .withComponent<ActiveEffectsComponent>('active-effects', { effects: [] }).entity
}

export type CharacterCreatorType = keyof typeof characterCreatorsDefinition
export const characterCreators: { [key in CharacterCreatorType]: CharacterCreator } = characterCreatorsDefinition
