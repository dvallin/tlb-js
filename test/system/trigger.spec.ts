import { registerComponents, TlbWorld } from '../../src/tlb'
import { Entity } from '../../src/ecs/entity'
import { World } from '../../src/ecs/world'
import { characterCreators } from '../../src/assets/characters'
import { Trigger } from '../../src/systems/trigger'
import { Uniform } from '../../src/random/distributions'
import { Random } from '../../src/random'
import { addDialog, AnswerType } from '../../src/assets/dialogs'
import { mockUi, mockReturnValue, callArgument, mockLog } from '../mocks'
import { UI } from '../../src/resources/ui'
import { State } from '../../src/game-states/state'
import { TriggeredByComponent } from '../../src/components/trigger'
import { createAssetFromShape } from '../../src/components/asset'
import { Rectangle } from '../../src/geometry/rectangle'
import { WorldMapResource, WorldMap } from '../../src/resources/world-map'
import { createFeatureFromType } from '../../src/components/feature'
import { Log } from '../../src/resources/log'
import { renderMap } from '../render'

describe('Trigger', () => {
  let world: TlbWorld
  let player: Entity
  let guard: Entity
  let ui: UI
  let log: Log
  let map: WorldMap
  let pushState: () => void
  let system: Trigger
  beforeEach(() => {
    world = new World()
    registerComponents(world)
    world.registerResource(new WorldMapResource(4))
    map = world.getResource<WorldMapResource>('map')
    ui = mockUi(world)
    log = mockLog(world)

    player = characterCreators.player(world)
    guard = characterCreators.guard(world)
    world.editEntity(player).withComponent('start-turn', {})
    pushState = jest.fn()

    system = new Trigger(new Random(new Uniform('1')), pushState)
  })

  describe('dialog triggers', () => {
    it('opens a dialog modal', () => {
      addDialog(world, guard, 'randomRemarks')
      world.editEntity(guard).withComponent<TriggeredByComponent>('triggered-by', { entity: player })

      system.update(world, guard)

      expect(ui.showDialogModal).toHaveBeenCalled()
      expect((callArgument(pushState, 0, 0) as State).name).toEqual('modal')
    })

    it('closes dialog', () => {
      mockReturnValue<boolean>(ui.dialogShowing, true)
      mockReturnValue<AnswerType>(ui.dialogResult, 'close')
      addDialog(world, guard, 'randomRemarks')
      world.editEntity(guard).withComponent<TriggeredByComponent>('triggered-by', { entity: player })

      system.update(world, guard)

      expect(world.hasComponent(guard, 'active')).toBeFalsy()
      expect(world.hasComponent(guard, 'triggered-by')).toBeFalsy()
    })

    it('handles attack result', () => {
      mockReturnValue<boolean>(ui.dialogShowing, true)
      mockReturnValue<AnswerType>(ui.dialogResult, 'attack')
      addDialog(world, guard, 'randomRemarks')
      world.editEntity(guard).withComponent<TriggeredByComponent>('triggered-by', { entity: player })

      system.update(world, guard)

      expect(world.hasComponent(guard, 'wait-turn')).toBeTruthy()
      expect(world.hasComponent(player, 'start-turn')).toBeTruthy()
    })
  })

  describe('asset triggers', () => {
    describe('door', () => {
      it('opens', () => {
        map.levels[0].boundary.foreach(p => createFeatureFromType(world, 0, p, 'corridor'))
        const door = createAssetFromShape(world, 0, new Rectangle(1, 1, 3, 1), 'door')

        world.editEntity(door).withComponent<TriggeredByComponent>('triggered-by', { entity: player })
        system.update(world, door)

        expect(renderMap(world)).toMatchSnapshot()
      })

      it('opens and then closes', () => {
        map.levels[0].boundary.foreach(p => createFeatureFromType(world, 0, p, 'corridor'))
        const door = createAssetFromShape(world, 0, new Rectangle(1, 1, 3, 1), 'door')

        world.editEntity(door).withComponent<TriggeredByComponent>('triggered-by', { entity: player })
        system.update(world, door)
        world.editEntity(door).withComponent<TriggeredByComponent>('triggered-by', { entity: player })
        system.update(world, door)

        expect(renderMap(world)).toMatchSnapshot()
      })
    })

    describe('loot', () => {
      it('lootable opens dialog and pushes modal state', () => {
        map.levels[0].boundary.foreach(p => createFeatureFromType(world, 0, p, 'corridor'))
        const loot = createAssetFromShape(world, 0, new Rectangle(1, 1, 1, 1), 'loot')

        world.editEntity(loot).withComponent<TriggeredByComponent>('triggered-by', { entity: player })
        system.update(world, loot)

        expect(ui.showInventoryTransferModal).toHaveBeenCalled()
        expect((callArgument(pushState, 0, 0) as State).name).toEqual('modal')
      })

      it('removes lootable if it is empty after close', () => {
        ui.isModal = true
        mockReturnValue<boolean>(ui.inventoryTransferModalShowing, false)
        map.levels[0].boundary.foreach(p => createFeatureFromType(world, 0, p, 'corridor'))
        const loot = createAssetFromShape(world, 0, new Rectangle(1, 1, 1, 1), 'loot')

        world.editEntity(loot).withComponent<TriggeredByComponent>('triggered-by', { entity: player })
        system.update(world, loot)

        expect(ui.isModal).toBeFalsy()
        expect(world.hasEntity(loot)).toBeFalsy()
      })
    })

    describe('container', () => {
      it('lootable opens dialog and pushes modal state', () => {
        map.levels[0].boundary.foreach(p => createFeatureFromType(world, 0, p, 'corridor'))
        const loot = createAssetFromShape(world, 0, new Rectangle(1, 1, 1, 1), 'table')

        world.editEntity(loot).withComponent<TriggeredByComponent>('triggered-by', { entity: player })
        system.update(world, loot)

        expect(ui.showInventoryTransferModal).toHaveBeenCalled()
        expect((callArgument(pushState, 0, 0) as State).name).toEqual('modal')
      })

      it('does not remove lootable if it is empty after close', () => {
        ui.isModal = true
        mockReturnValue<boolean>(ui.inventoryTransferModalShowing, false)
        map.levels[0].boundary.foreach(p => createFeatureFromType(world, 0, p, 'corridor'))
        const loot = createAssetFromShape(world, 0, new Rectangle(1, 1, 1, 1), 'trash')

        world.editEntity(loot).withComponent<TriggeredByComponent>('triggered-by', { entity: player })
        system.update(world, loot)

        expect(ui.isModal).toBeFalsy()
        expect(world.hasEntity(loot)).toBeTruthy()
      })
    })

    describe('generator', () => {
      it('generators log a noise', () => {
        const shape = new Rectangle(0, 0, 2, 2)
        shape.foreach(p => createFeatureFromType(world, 0, p, 'corridor'))
        const generator = createAssetFromShape(world, 0, shape, 'generator')

        world.editEntity(generator).withComponent<TriggeredByComponent>('triggered-by', { entity: player })
        system.update(world, generator)

        expect(log.text).toHaveBeenCalledWith('clonck')
      })
    })
  })
})
