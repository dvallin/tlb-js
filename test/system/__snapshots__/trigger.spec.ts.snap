// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trigger asset triggers door opens 1`] = `
"....
....
....
...."
`;

exports[`Trigger asset triggers door opens and then closes 1`] = `
"....
.◊◊◊
....
...."
`;
